### PR TITLE
Health as a growing stat and health regen added.

### PR DESCRIPTION
**Pull request by ChrisDaBang**

### CODE REVIEW

File: Ball of Duty Server/Domain/Entities/Character.cs [hunk 1]
```diff
@@ -17,17 +17,19 @@ public class Character : GameObject, ICollidable
...
         private const double ALLOWED_SCORE_BEFORE_DECAY = 400;
+        private int _healthRegen;
         private int _killCount = 0;
-        private const long SCORE_DECAY_INTERVAL = 5000;
-        private readonly LightEvent _decayScoreEvent;
+        private const long Decay_And_Regen_INTERVAL = 5000;
+        private readonly LightEvent _decayAndRegenEvent;
 
-        public Character(double size, int health, Specializations specialization)
+        public Character(double size, int health, Specializations specialization, int HealthREGEN)
```
Comment (LuqJensen) [inline Ball of Duty Server/Domain/Entities/Character.cs]: wtf...

File: Ball of Duty Server/Domain/Entities/Character.cs [hunk 2]
```diff
@@ -17,17 +17,19 @@ public class Character : GameObject, ICollidable
         private const double SCORE_UP_FACTOR = 0.1;
         private const double SCORE_DECAY_FACTOR = 0.01;
         private const double ALLOWED_SCORE_BEFORE_DECAY = 400;
+        private int _healthRegen;
         private int _killCount = 0;
-        private const long SCORE_DECAY_INTERVAL = 5000;
-        private readonly LightEvent _decayScoreEvent;
+        private const long Decay_And_Regen_INTERVAL = 5000;
```
Comment (LuqJensen) [inline Ball of Duty Server/Domain/Entities/Character.cs]: wtf...

File: Ball of Duty Server/Domain/Entities/Character.cs [hunk 3]
```diff
@@ -79,5 +97,7 @@ public override bool Destroy(GameObject exterminator)
             }
             return false;
         }
+
+        public abstract void UpdateServerStats();
```
Comment (LuqJensen) [inline Ball of Duty Server/Domain/Entities/Character.cs]: ServerStats??
Whats wrong with UpdateStats?

File: Ball of Duty Server/Domain/Entities/Character.cs [hunk 4]
```diff
@@ -45,18 +47,34 @@ public void AddKill(Character victim)
...
 
+        public void DecayAndRegen()
+        {
+            DecayScore();
+            RegenHealth();
+        }
+
+        public void RegenHealth()
+        {
+            if (Health.Value < Health.Max)
```
Comment (LuqJensen) [inline Ball of Duty Server/Domain/Entities/Character.cs:58]: what if you have 1 health less than max, then you will suddenly have more health than max when this method executes..

File: Ball of Duty Server/Domain/Entities/Character.cs [hunk 5]
```diff
@@ -45,18 +47,34 @@ public void AddKill(Character victim)
             {
                 HighScore = Score;
             }
+            UpdateServerStats();
             NotifyObservers(Observation.ACQUISITION_OF_GOLD, victim);
         }
 
+        public void DecayAndRegen()
```
Comment (LuqJensen) [inline Ball of Duty Server/Domain/Entities/Character.cs]: LightEvent is VERY cheap. you really should keep logic like this separate. make one event for each, that will also futureproof the design in the case that decay and regen shouldnt happen at the same interval

File: Ball of Duty Server/Domain/Entities/Character.cs [hunk 6]
```diff
@@ -17,17 +17,19 @@ public class Character : GameObject, ICollidable
         private const double SCORE_UP_FACTOR = 0.1;
         private const double SCORE_DECAY_FACTOR = 0.01;
         private const double ALLOWED_SCORE_BEFORE_DECAY = 400;
+        private int _healthRegen;
```
Comment (LuqJensen) [inline Ball of Duty Server/Domain/Entities/Character.cs]: shouldnt this be part of the Health class?

File: Ball of Duty Server/Domain/Entities/CharacterSpecializations/Blaster.cs [hunk 7]
```diff
@@ -6,9 +7,20 @@ public class Blaster : Character
...
 
-        public Blaster() : base(SIZE, HP, Specializations.BLASTER)
+        public Blaster() : base(SIZE, HP, Specializations.BLASTER, HP_REGEN)
         {
         }
+
+        public override void UpdateServerStats()
+        {
+            Health.Max = HP + (int)(Score * HP_INCREASE_FACTOR);
+            if (Health.Value > Health.Max)
```
Comment (LuqJensen) [inline Ball of Duty Server/Domain/Entities/CharacterSpecializations/Blaster.cs]: this is useless. you call UpdateServerStats through DecayScore before RegenHealth is called.
that means that despite this check Health.Value may still exceed Health.Max

File: Ball of Duty Server/Domain/Entities/Character.cs (hunk 3 above)
Comment (ChrisDaBang) [inline Ball of Duty Server/Domain/Entities/Character.cs]: Well, just wanted to point out it's for updating the stats the server is responsible for, it doesn't call to the clients for them to update stats

Comment (LuqJensen) [inline Ball of Duty Server/Domain/Entities/Character.cs]: pls name it UpdateStats. Character will never be responsible for updating anything clientside either way

**Pull request closed without merging**